calico/common.py

```diff
@@ -601,105 +608,132 @@ def _validate_rules(rules_dict, issues):
+        if tag is None:
+            continue
+        if not VALID_ID_RE.match(tag):
+            issues.append("Invalid %s: %r." % (tag_type, tag))
+
+    # For selectors, we replace the value with the parsed selector.
+    # This avoids having to re-parse it later and it ensures that
+    # equivalent selectors compare equal.
+    for sel_type in (neg_pfx + 'src_selector', neg_pfx + 'dst_selector'):
+        sel_str = rule.get(sel_type)
+        if sel_str is None:
+            # sel_type wasn't present.
+            continue
+        try:
+            sel = parse_selector(sel_str)
+        except BadSelector:
+            issues.append("Invalid %s: %r" % (sel_type, sel_str))
+        else:
+            rule[sel_type] = sel
+
+    for key in (neg_pfx + "src_net", neg_pfx + "dst_net"):
+        network = rule.get(key)
+        if (network is not None and
+            not validate_cidr(rule[key], ip_version)):
+            issues.append("Invalid CIDR (version %s) in rule %s." %
+                          (ip_version, rule))
+        elif network is not None:
+            rule[key] = canonicalise_cidr(network, ip_version)
+    for key in (neg_pfx + "src_ports", neg_pfx + "dst_ports"):
+        ports = rule.get(key)
+        if (ports is not None and
+            not isinstance(ports, list)):
+            issues.append("Expected ports to be a list in rule %s."
+                          % rule)
+            continue
+
+        if ports is not None:
+            if protocol not in KERNEL_PORT_PROTOCOLS:
+                issues.append("%s is not allowed for protocol %s in "
+                              "rule %s" % (key, protocol, rule))
+            for port in ports:
+                error = validate_rule_port(port)
+                if error:
+                    issues.append("Invalid port %s (%s) in rule %s." %
+                                  (port, error, rule))
+
+    action = rule.get(neg_pfx + 'action')
+    if (action is not None and
+            action not in KNOWN_ACTIONS):
+        issues.append("Invalid action in rule %s." % rule)
+
+    icmp_type = rule.get(neg_pfx + 'icmp_type')
+    if icmp_type is not None:
+        if not isinstance(icmp_type, int):
+            issues.append("ICMP type is not an integer in rule %s." %
+                          rule)
+        elif not 0 <= icmp_type <= 255:
+            issues.append("ICMP type is out of range in rule %s." %
+                          rule)
+    icmp_code = rule.get(neg_pfx + "icmp_code")
+    if icmp_code is not None:
+        if not isinstance(icmp_code, int):
+            issues.append("ICMP code is not an integer in rule %s." %
+                          rule)
+        elif not 0 <= icmp_code <= 255:
+            issues.append("ICMP code is out of range.")
+        if icmp_type is None:
+            # ICMP code without ICMP type not supported by iptables;
+            # firewall against that.
+            issues.append("ICMP code specified without ICMP type.")
```

Comment:
Do you want `in rule %s` in these issue descriptions?

(Not actually reviewing this, I just glanced at the code and happened to spot this.)